---
title: "QC plots"
author: "Tom Smith"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
  pdf:
    theme: default
---

This notebook generates QC plots from the guide RNA (gRNA) counts table ('quant.dir/all_samples.tsv').
To see the code, click the 'code' buttons, or set 'show all code' using the 'code' drop down in the top-right.

```{r Load libraries, message=FALSE, warning=FALSE}
library(here)

library(tidyr)
library(dplyr)
library(ggplot2)

library(DescTools)
library(pheatmap)
library(DESeq2)
```

```{r Define plot functions}
# Defining a couple of functions which we will use later in the plotting
get_my_colours <- function (n) 
{
    if ((n > 12 | n < 1)) 
        stop("n must be >=1 & <=12")
    cbPalette7 <- c("#2271B2", "#d55e00", "#359B73", "#e69f00", 
        "#3DB7E9", "#f0e442", "#F748A5")
    cbPalette12 <- c("#9F0162", "#009F81", "#FF5AAF", "#00FCCF", 
        "#8400CD", "#008DF9", "#00C2F9", "#FFB2FD", "#A40122", 
        "#E20134", "#FF6E3A", "#FFC33B")
    if (n <= 7) {
        return(cbPalette7[1:n])
    }
    else {
        return(cbPalette12[1:n])
    }
}
  
my_theme <- function (base_size = 20, base_family = "serif", aspect_square = TRUE, 
    border = TRUE) 
{
    camprot_theme <- theme_bw(base_size = base_size, base_family = base_family) + 
        theme(panel.grid = element_blank(), panel.background = element_blank())
    if (aspect_square) {
        camprot_theme <- camprot_theme + theme(aspect.ratio = 1)
    }
    if (!border) {
        camprot_theme <- camprot_theme + theme(panel.border = element_blank(), 
            axis.line = element_line(colour = "black", size = 0.5))
    }
}
```


```{r Read in input}
# Read in the input
gRNA_counts <- read.table(here('quant.dir/all_samples.tsv'), header=TRUE) %>%
  filter(sgRNA!='*')

if('Undetermined' %in% colnames(gRNA_counts)){
  gRNA_counts <- gRNA_counts %>% select(-Undetermined)
}

gRNA_counts_exp_only <- gRNA_counts %>% 
  tibble::column_to_rownames('sgRNA') %>% 
  select(-gene) %>%
  as.matrix()

error_counts <- read.table(here('bowtie.dir/all_errors.tsv'), sep='\t', header=TRUE)
error_counts$errors[is.na(error_counts$errors)] <- 'Unmapped'


experimental_design <- read.delim(here('experimental_design.tsv'), sep='\t', row.names=1)


n_guides <- length(readLines(file(here('guides.csv')), warn=FALSE))
```


# Errors
```{r plot error tallies, message=FALSE}
max_errors <- setdiff(error_counts$errors, 'Unmapped') %>% as.numeric() %>% max()

to_plot <- error_counts %>% pivot_longer(cols=-errors, names_to='sample', values_to='n') %>%
  merge(experimental_design, by.x='sample', by.y='row.names') %>%
  mutate(errors=factor(errors, levels=c(0:max_errors, 'Unmapped'))) %>%
  group_by(sample) %>%
  mutate(perc=100*n/sum(n))

p <- to_plot %>%
  ggplot(aes(sample, perc, fill=errors)) +
  geom_bar(stat='identity', position='stack') +
  scale_fill_manual(values=c(get_my_colours(max_errors+1), 'grey'), name='Errors') +
  geom_text(data=(filter(to_plot, errors %in% c('0', 'Unmapped'))), aes(label=round(perc)), position=position_stack(vjust = .5)) +
  xlab('') +
  ylab('Percentage of reads') +
  ggtitle('Number of errors in reads\nrelative to guide RNA sequences') +
  my_theme(border=FALSE, base_family='sans', base_size=8) +
  theme(axis.text.x=element_text(angle=45, vjust=1, hjust=1))
  
print(p)

ggsave(here('qc_plots.dir/error_tallies.pdf'))
ggsave(here('qc_plots.dir/error_tallies.png'))  

```


# Gini coefficient

The [Gini coefficient](https://en.wikipedia.org/wiki/Gini_coefficient) is a useful summary of how equal the distribution of gRNA counts are, where 0 = perfect equality and 1 = perfect inequality.

For a CRISPR screen, we expect unsorted and plasmid library samples to have the smallest Gini coefficients.

```{r plot Gini coefficient, message=FALSE}
p <- gRNA_counts_exp_only %>% apply(MARGIN=2, Gini) %>% data.frame() %>%
  tibble::rownames_to_column('sample') %>%
  dplyr::rename('gini'='.') %>%
  ggplot(aes(sample, gini)) +
  geom_bar(stat='identity') +
  my_theme(base_size=10, border=FALSE) +
  theme(axis.text.x=element_text(angle=45, vjust=1, hjust=1)) +
  xlab('')+
  ylab('Gini coefficient')  

print(p)

ggsave(here('qc_plots.dir/gini_coefficients.pdf'))
ggsave(here('qc_plots.dir/gini_coefficients.png'))
```


# gRNAs per gene

The number of guide RNAs per gene. This includes all detected gRNAs across the samples (e.g > 0 reads in any sample).

```{r Plot gRNA per gene, message=FALSE}
# remove gRNA with zero counts 
gRNA_counts_noMissing <- gRNA_counts[rowSums(gRNA_counts_exp_only)>0,]

n_gRNA_per_gene_tally <- gRNA_counts_noMissing %>%
  pull(gene) %>%
  table() %>%
  table() %>%
  data.frame() %>%
  setNames(c('n_gRNAs', 'genes'))

p <- n_gRNA_per_gene_tally %>%
  ggplot(aes(n_gRNAs, genes)) +
  geom_bar(stat='identity') +
  geom_text(aes(label=genes), hjust=-0.25, angle=90, size=3) +
  ylim(c(NA, max(n_gRNA_per_gene_tally$genes*1.2))) +
  my_theme(border=FALSE, base_family='sans', base_size=12) +
  theme(axis.text.x=element_text(size=8, angle=45, vjust=1, hjust=1)) +
  xlab('Number of gRNAs') +
  ylab('Genes')

print(p)

ggsave(here('qc_plots.dir/gRNAs_per_gene.pdf'))
ggsave(here('qc_plots.dir/gRNAs_per_gene.png'))
```


# gRNA coverage (how many & at what depth) for each sample

Grey-dashed horizontal line is the total number of gRNAs in the pool, inluding those not detected.
```{r plot guides per sample, message=FALSE}

guides_per_sample <- colSums((select(gRNA_counts, -c(gene, sgRNA)))>0) %>%
  data.frame() %>%
  setNames('gRNAs') %>%
  merge(experimental_design, by='row.names') %>%
  rename('Row.names'='Sample')
  
p <- guides_per_sample %>%
  ggplot(aes(factor(Sample), as.numeric(gRNAs))) +
  geom_bar(stat='identity') +
  my_theme(border=FALSE, base_size=15, base_family='sans') +
  theme(axis.text.x=element_text(angle=45, vjust=1, hjust=1)) +
  xlab('') +
  geom_text(aes(label=gRNAs), hjust=-0.25, angle=90, size=3) +
  ylim(c(NA, max(guides_per_sample$gRNAs*1.15))) +
  geom_hline(yintercept=n_guides, linetype=2, colour=get_my_colours(1), size=0.25) +
  ylab('Guides') 

ggsave(here('qc_plots.dir/gRNAs_per_sample.pdf'))
ggsave(here('qc_plots.dir/gRNAs_per_sample.png'))

```

```{r plot read depth per guide, message=FALSE}
p <- gRNA_counts %>%
  pivot_longer(cols=-c(gene, sgRNA)) %>%
  merge(experimental_design, by.x='name', by.y='row.names') %>%
  filter(!is.na(condition)) %>%
  ggplot(aes(value, group=name, colour=sort, linetype=condition)) +
  geom_density() +
  scale_x_log10() +
  my_theme(border=FALSE, base_size=15, base_family='sans') +
  xlab('Reads') +
  ylab('density') +
  scale_colour_manual(values=get_my_colours(3)) +
  ggtitle('Reads per gRNA')

print(p)

ggsave(here('qc_plots.dir/gRNAs_per_sample.pdf'))
ggsave(here('qc_plots.dir/gRNAs_per_sample.png'))

```

# Was Sequencing depth sufficient/appropriate?

By re-sampling from the counts table at different depths, we can simulate the impact of lower or higher sequencing depth and see how this would affect the number of gRNAs detecting at different reads thresholds. 
```{r count guides in resampled count tables, message=FALSE}
resampled_quant <- Sys.glob(here('resampled_quant.dir/*all_samples.tsv'))

all_resampled_guides_per_sample <- lapply(resampled_quant, function(infile){
  depth <- strsplit(basename(infile), '_',)[[1]][[1]]
  resampled_gRNA_counts <- read.table(infile, header=TRUE) %>%
  filter(sgRNA!='*')
  
  if('Undetermined' %in% colnames(resampled_gRNA_counts)){
    resampled_gRNA_counts <- resampled_gRNA_counts %>% select(-Undetermined)
   }
  
  resampled_guides_per_sample <- c(1,10,100) %>% lapply(function(threshold){
    colSums((select(resampled_gRNA_counts, -c(gene, sgRNA)))>=threshold) %>%
      data.frame() %>%
      setNames('gRNAs') %>%
      mutate(reads_threshold=threshold, depth=depth) %>%
      tibble::rownames_to_column('sample')
  }) %>%
    bind_rows()
  
}) %>%
  bind_rows()
```

Grey-dashed horizontal line is the total number of gRNAs in the pool, including those not detected.
```{r plot guides counts in resampled count tables, message=FALSE}
p <- all_resampled_guides_per_sample %>%
  mutate(reads_threshold=sprintf('>%s reads', reads_threshold)) %>%
  merge(experimental_design, by.x='sample', by.y='row.names') %>%
  ggplot(aes(as.numeric(depth), gRNAs, group=sample, colour=sort, linetype=condition)) +
  geom_line() +
  facet_grid(reads_threshold~.) +
  my_theme(border=FALSE, base_family='sans', base_size=15) +
  theme(strip.background=element_blank()) +
  geom_vline(xintercept=1, linetype=2, colour='grey') +
  geom_hline(yintercept=n_guides, linetype=2, colour='grey') +
  xlab('Sequencing depth\n(relative to actual depth)')

print(p)


ggsave(here('qc_plots.dir/sequencing_depth.pdf'))
ggsave(here('qc_plots.dir/sequencing_depth.png'))
```


```{r}
cts <- gRNA_counts_exp_only[rowSums(gRNA_counts_exp_only)>0,]

```

# Correlation matrix

Visualising the correlations between samples can be useful to identify outlier samples and, where there are replicate experiments, check that they have simiar guide RNA counts.

```{r plot correlation matrix, fig.height=7, fig.width=7, message=FALSE}

mat <- cor(cts)
pheatmap(mat, cellwidth=10, cellheight=10)
pheatmap(mat, cellwidth=10, cellheight=10, filename=here('qc_plots.dir/correlations.pdf'))
pheatmap(mat, cellwidth=10, cellheight=10, filename=here('qc_plots.dir/correlations.png'))
```
# PCA

Principle Component Analysis is useful to identify the major sources of variance. We expect the major sources of variance to align with experimental conditions. 

```{r plot PCA projections, message=FALSE}

experimental_design_complete <- experimental_design[experimental_design$sort!='Unsorted',]

dds <- DESeqDataSetFromMatrix(countData = cts[,rownames(experimental_design_complete)],
                              colData = experimental_design_complete,
                              design = as.formula(sprintf('~ %s', paste(colnames(experimental_design_complete), collapse='+'))))

vst <- vst(dds, blind=FALSE)

p <- DESeq2::plotPCA(vst, intgroup=colnames(experimental_design_complete), ntop=nrow(vst)) +
  my_theme(border=FALSE, base_family='sans') +
  aes(fill=condition, shape=sort, colour=condition) +
  geom_point(colour='grey30', size=3) +
  scale_fill_manual(values=get_my_colours(2)) +
  scale_shape_manual(values=21:25) +
  scale_colour_discrete(guide="none") +
  guides(fill = guide_legend(override.aes = list(pch=21) ) )

print(p)

ggsave(here('qc_plots.dir/pca.png'))
ggsave(here('qc_plots.dir/pca.pdf'))
```



